[all] Split unit rule tests for rules into single classes

adangel opened the pull request:
When working on a single rule, I find it more convenient, to be able to run just the tests for the specific rule instead of all tests for the complete ruleset.

I've split java's best practices ruleset into single unit test classes as an example.

What do you think?

If you agree, I'll do the same for all the other ruleset/rule tests.

oowekyala commented on pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/bestpractices/CheckResultSetTest.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public void setUp() {
+        addRule(RULESET, "CheckResultSet");
```
Some implementation thoughts:

We could use convention the existing naming conventions to infer the name of the rule and the category reference.
* The rule name can be determined from the class name: `getClass().getSimpleName().replaceFirst("Test$", "")`
* The RULESET reference is also easy to infer from our package naming conventions, eg by matching the package name with the following regex: `.*lang\.(\w+?)\.rule\.(\w+)$`.

This would obviate the need for 
* having a base class for each category
* implementing anything in concrete rule test classes

If we migrate tests of all rulesets to that new format (single class per rule), then I think we could use another abstract class than SimpleAggregatorTst (since we don't aggregate anything anymore), e.g. ConventionalSingleRuleTest (internal api I guess, since it relies on our naming conventions), that would contain the implementation

Sidenote, in case we start using Kotlin to test rules, we could even avoid having a file for each test and just declare a bunch of classes in a single file (per category), e.g.:
```kotlin
class CheckResultSetTest : ConventionalSingleRuleTest()
class ConstantsInInterfaceTest : ConventionalSingleRuleTest()
...
```
which can be run individually

adangel replied:
Yes, we can use convention for the rule name. I think, I'll change that, it should work.
Using convention for the ruleset should work too, and I like the idea to have a separate base test class for our conventional based rule tests...
I'll try to update https://pmd.github.io/pmd-6.6.0/pmd_userdocs_extending_testing.html , too.

About naming of SimpleAggregatorTst: That's argueable - we still aggregate multiple tests into one test class :smile: But we could think of a better name for PMD 7 - maybe BaseRuleTest? It would be part of the public API to support unit testing custom rules.

adangel commented:
@oowekyala @jsotuyod 
I've updated this PR and changed the rule tests for all languages now.

Most of the test classes look now like this: https://github.com/adangel/pmd/blob/unit-test-split/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AbstractClassWithoutAbstractMethodTest.java

When there were unit tests, I've merged them into the correct rule test class, e.g. https://github.com/adangel/pmd/blob/unit-test-split/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/design/StdCyclomaticComplexityTest.java

The same with helper classes for typeresolution, e.g. https://github.com/adangel/pmd/blob/unit-test-split/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryFullyQualifiedNameTest.java

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: Please check the <a href="http...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Please check the <a href="http://chunk.io/pmd/651f0ecac0c3495aa154efd7a3c86101/diff/index.html">regression diff report</a> to make sure that everything is expected</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

oowekyala commented on pmd-test/src/main/java/net/sourceforge/pmd/testframework/SimpleAggregatorTst.java:
```diff
@@ -20,13 +20,20 @@
     protected void setUp() {
-        // empty, to be overridden
+        String[] packages = getClass().getPackage().getName().split("\\.");
+        String categoryName = packages[packages.length - 1];
```
This is PMD specific, I guess it would probably fail most of the time outside of PMD. Wouldn't it be better to use our own subclass of SimpleAggregatorTst (or RuleTst, since we test a single rule)?

adangel replied:
That's a good point. The default implementation will probably only work for PMD itself, hence in the documentation this method is always overridden, if you create your own tests.
I'll see, what I can do about this...

oowekyala commented:
Looks good to me. I'll merge it in the evening if no more comments arise

oowekyala commented:
And thanks @adangel for taking this on this tedious task !
